core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -237,19 +245,12 @@ protected void doStart() {
     private void tryStart(Profiler profiler) {
         try {
             String dockerImageName = image.get();
-            logger().debug("Starting container: {}", dockerImageName);
-
-            logger().info("Creating container for image: {}", dockerImageName);
-            profiler.start("Create container");
-            CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
-            applyConfiguration(createCommand);
-
-            containerId = createCommand.exec().getId();
-            copyToFileContainerPathMap.forEach(this::copyFileToContainer);
 
-            logger().info("Starting container with ID: {}", containerId);
-            profiler.start("Start container");
-            dockerClient.startContainerCmd(containerId).exec();
+            if (isReusingEnabledWhenReusable) {
+                tryUseReusableContainer(profiler, dockerImageName);
```

Comment:
I would extracted tryUseReusableContainer() and tryStartDisposableContainer() methods into separate classes like ReusableContainerLaunchStrategy and also methods createContainer() (with applyConfiguration private method) and startContainer() to Command pattern classes but it would require moving a lot of existing code from GenericContainer so to keep diff as small as possible I wrote new code in a procedure way like the remaining code of GenericContainer.